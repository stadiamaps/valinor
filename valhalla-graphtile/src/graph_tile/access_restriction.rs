use crate::graph_tile::GraphTileError;
use crate::Access;
use bitfield_struct::bitfield;
use enumset::EnumSet;
use zerocopy::try_transmute;
use zerocopy_derive::{FromBytes, TryFromBytes};

/// Types of access restrictions.
#[derive(TryFromBytes, Debug, Eq, PartialEq)]
#[repr(u8)]
pub enum AccessRestrictionType {
    Hazmat,
    MaxHeight,
    MaxWidth,
    MaxLength,
    MaxWeight,
    MaxAxleLoad,
    TimedAllowed,
    TimedDenied,
    DestinationAllowed,
    MaxAxles,
    // Valhalla docs say the max supported value is 31,
    // but the field looks 6 bits wide to me, not 5.
}

#[bitfield(u64)]
#[derive(PartialEq, Eq, FromBytes)]
struct AccessRestrictionBitField {
    #[bits(22)]
    edge_index: u32,
    #[bits(6)]
    restriction_type: u8,
    #[bits(12)]
    modes: u16,
    #[bits(24)]
    _spare: u32,
}

/// Access restrictions beyond the usual access tags
#[derive(PartialEq, Eq, TryFromBytes, Debug)]
#[repr(C)]
pub struct AccessRestriction {
    bitfield: AccessRestrictionBitField,
    value: u64,
}

impl AccessRestriction {
    /// Gets the edge index (within the tile) to which the restriction applies.
    #[inline]
    pub fn edge_index(&self) -> u32 {
        self.bitfield.edge_index()
    }

    /// Gets the type of access restriction.
    ///
    /// # Panics
    ///
    /// This function currently panics if the edge use contains an invalid bit pattern.
    /// We should check the validity of the field when we do the initial try_transmute
    /// on the type (and then have an error result that the tile is invalid).
    /// There are some upstream bugs though between zerocopy and bitfield-struct
    /// macros: https://github.com/google/zerocopy/issues/388.
    #[inline]
    pub fn restriction_type(&self) -> AccessRestrictionType {
        try_transmute!(self.bitfield.restriction_type())
            .map_err(|_| GraphTileError::ValidityError)
            .expect("Invalid bit pattern for restriction type.")
    }

    /// The access modes affected by this restriction.
    #[inline]
    pub fn affected_access_modes(&self) -> EnumSet<Access> {
        // TODO: Look at ways to do this with FromBytes; this currently copies
        // Safety: The access bits are length 12, so invalid representations are impossible.
        unsafe { EnumSet::from_repr_unchecked(self.bitfield.modes()) }
    }
}

#[cfg(test)]
mod tests {
    use super::{Access, AccessRestrictionType};
    use crate::graph_tile::TEST_GRAPH_TILE;
    use enumset::EnumSet;

    #[test]
    fn test_parse_access_restrictions_count() {
        let tile = &*TEST_GRAPH_TILE;

        assert_eq!(
            tile.access_restrictions.len(),
            tile.header.access_restriction_count() as usize
        );
    }

    #[test]
    fn test_parse_access_restrictions() {
        let tile = &*TEST_GRAPH_TILE;

        insta::assert_debug_snapshot!(tile.access_restrictions[0]);

        assert_eq!(
            tile.access_restrictions[0].restriction_type(),
            AccessRestrictionType::MaxWeight
        );
        assert_eq!(
            tile.access_restrictions[0].affected_access_modes(),
            EnumSet::from(Access::Truck)
        );

        insta::assert_debug_snapshot!(tile.access_restrictions.last().unwrap());

        // TODO: Other sanity checks after we add some more advanced methods
    }
}
