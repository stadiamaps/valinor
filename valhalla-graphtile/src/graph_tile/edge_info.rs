use crate::{
    graph_tile::GraphTileError, shape_codec::decode_shape, transmute_variable_length_data, AsCowStr,
};
use bitfield_struct::bitfield;
use bytes::Bytes;
use bytes_varint::VarIntError;
use geo::LineString;
use std::borrow::Cow;
use std::cell::OnceCell;
use zerocopy::transmute;
use zerocopy_derive::FromBytes;

#[derive(FromBytes)]
#[bitfield(u32)]
pub struct NameInfo {
    #[bits(24)]
    name_offset: u32,
    /// Additional fields following the name.
    ///
    /// These can be used for additional information like phonetic readings.
    #[bits(4)]
    additional_fields: u8,
    #[bits(1)]
    is_route_num: u8,
    /// Indicates the string is specially tagged (ex: uses the first char as the tag type).
    ///
    /// This doesn't have any relation to OSM tagging.
    #[bits(1)]
    is_tagged: u8,
    #[bits(2)]
    _spare: u8,
}

#[derive(FromBytes)]
#[bitfield(u32)]
struct FirstInnerBitfield {
    #[bits(12)]
    mean_elevation: u16,
    #[bits(4)]
    bike_network: u8,
    #[bits(8)]
    speed_limit: u8,
    #[bits(8)]
    extended_way_id: u8,
}

#[derive(FromBytes)]
#[bitfield(u32)]
struct SecondInnerBitfield {
    #[bits(4)]
    name_count: u8,
    #[bits(16)]
    encoded_shape_size: u16,
    #[bits(8)]
    extended_way_id: u8,
    #[bits(2)]
    extended_way_id_size: u8,
    #[bits(1)]
    has_elevation: u8,
    #[bits(1)]
    _spare: u8,
}

#[derive(Debug, FromBytes)]
#[repr(C)]
struct EdgeInfoInner {
    // The first part of the OSM way ID
    // TODO: would be nice to reshuffle this in v4.0
    way_id: u32,
    first_inner_bitfield: FirstInnerBitfield,
    second_inner_bitfield: SecondInnerBitfield,
}

#[derive(Debug)]
#[repr(C)]
pub struct EdgeInfo {
    inner: EdgeInfoInner,
    name_info_list: Vec<NameInfo>,
    /// The raw varint-encoded shape bytes.
    pub encoded_shape: Bytes,
    decoded_shape: OnceCell<LineString<f64>>,
    // TODO: Final 2 bytes of a 64-bit way ID
    // TODO: Encoded elevation (pointer?)
    text_list_memory: Bytes,
    // TODO: Tag cache
}

impl EdgeInfo {
    /// Gets the shape of the edge geometry as a LineString.
    ///
    /// # Performance
    ///
    /// If this [`EdgeInfo`] geometry has not been accessed previously,
    /// then it will be decoded during this method call.
    /// Subsequent fetches will be cached for as long as the `EdgeInfo`
    /// is live.
    pub fn shape(&self) -> Result<&LineString<f64>, VarIntError> {
        // TODO: Use https://doc.rust-lang.org/core/cell/struct.OnceCell.html#method.get_or_try_init when stabilized
        match self.decoded_shape.get() {
            Some(linestring) => Ok(linestring),
            None => {
                let shape = decode_shape(self.encoded_shape.clone())?;
                Ok(self.decoded_shape.get_or_init(|| shape))
            }
        }
    }

    // TODO: Other filters (tagged and linguistic filters)
    /// Gets all names for this edge.
    ///
    /// # Performance
    ///
    /// This is mostly just pointer indirection and some light filtering.
    /// Not great to call in a hot loop, but also not doing a lot of heavy processing.
    /// The main thing to be careful of in hot paths is allocations.
    pub fn get_names(&self) -> Vec<Cow<'_, str>> {
        self.name_info_list
            .iter()
            .filter_map(|ni| {
                // FIXME: Methods
                // No, this is not a bug...
                if ni.is_tagged() == 0 {
                    Some(self.text_list_memory[ni.name_offset() as usize..].as_cow_str())
                } else {
                    None
                }
            })
            .collect()
    }
}

// TODO: Feels like this could be a macro
impl TryFrom<(Bytes, Bytes)> for EdgeInfo {
    type Error = GraphTileError;

    fn try_from((bytes, text_list_memory): (Bytes, Bytes)) -> Result<Self, Self::Error> {
        let value = &bytes;
        const INNER_SIZE: usize = size_of::<EdgeInfoInner>();
        let inner_slice: [u8; INNER_SIZE] = (&value[0..INNER_SIZE]).try_into()?;
        let inner: EdgeInfoInner = transmute!(inner_slice);

        let (name_info_list, offset) = transmute_variable_length_data!(
            NameInfo,
            value,
            INNER_SIZE,
            inner.second_inner_bitfield.name_count() as usize
        )?;

        let encoded_shape =
            bytes.slice(offset..offset + inner.second_inner_bitfield.encoded_shape_size() as usize);

        Ok(Self {
            inner,
            name_info_list,
            encoded_shape,
            decoded_shape: OnceCell::new(),
            text_list_memory,
        })
    }
}
