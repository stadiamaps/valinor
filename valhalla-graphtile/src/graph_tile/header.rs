use crate::{AsCowStr, BIN_COUNT, GraphId};
use bitfield_struct::bitfield;
use geo::{Coord, coord};
use std::borrow::Cow;
use zerocopy_derive::FromBytes;

/// Remaining variable offset slots for growth.
/// See valhalla/balder/graphtileheader.h for details.
const EMPTY_SLOTS: usize = 11;

#[derive(FromBytes)]
#[bitfield(u64)]
struct FirstBitfield {
    #[bits(46)]
    graph_id: u64,
    #[bits(4)]
    density: u8,
    #[bits(4)]
    name_quality: u8,
    #[bits(4)]
    speed_quality: u8,
    #[bits(4)]
    exit_quality: u8,
    // NOTE: We don't use bool, because that doesn't have a `FromBytes` implementation.
    // We are only using a single bit though, so we *can* guarantee that the bit pattern is valid.
    // This will probably change once we have the ability to assume bit validity.
    // This is currently in beta: https://doc.rust-lang.org/beta/std/mem/trait.TransmuteFrom.html.
    #[bits(1)]
    has_elevation: u8,
    #[bits(1)]
    has_ext_directed_edge: u8,
}

#[derive(FromBytes)]
#[bitfield(u64)]
struct SecondBitfield {
    #[bits(21)]
    node_count: u32,
    #[bits(21)]
    directed_edge_count: u32,
    #[bits(21)]
    predicted_speeds_count: u32,
    #[bits(1)]
    _spare: u8,
}

// NOTE: There are a lot of weird paddings which arose because of an incompatibility in padding
// between x86 and amd64 architectures.
// This was entirely accidental upstream and will eventually be cleaned up in v4.

#[derive(FromBytes)]
#[bitfield(u32)]
struct TransitionCountBitfield {
    #[bits(22)]
    transition_count: u32,
    // Deprecated; scheduled for deletion in v4
    #[bits(10)]
    _spare: u16,
}

#[derive(FromBytes)]
#[bitfield(u32)]
struct TurnLaneCountBitfield {
    #[bits(21)]
    turn_lane_count: u32,
    #[bits(11)]
    _spare: u16,
}

#[derive(FromBytes)]
#[bitfield(u64)]
struct TransitRecordBitfield {
    #[bits(16)]
    transfer_count: u32,
    #[bits(7)]
    _spare: u16,
    #[bits(24)]
    departure_count: u32,
    #[bits(16)]
    stop_count: u16,
    #[bits(1)]
    _spare: u8,
}

#[derive(FromBytes)]
#[bitfield(u64)]
struct MiscCountsBitFieldOne {
    #[bits(12)]
    route_count: u16,
    #[bits(12)]
    schedule_count: u16,
    #[bits(24)]
    sign_count: u32,
    #[bits(16)]
    _spare: u16,
}

#[derive(FromBytes)]
#[bitfield(u64)]
struct MiscCountsBitFieldTwo {
    #[bits(24)]
    access_restriction_count: u32,
    #[bits(16)]
    admin_count: u16,
    #[bits(24)]
    _spare: u32,
}

// TODO: Unsure if packed is needed as well.
// Seems like it's technically not, since Valhalla is explicit about padding
/// Summary information about the graph tile.
///
/// This contains metadata like version,
/// number of nodes and edges,
/// and pointer offsets to other data.
#[derive(Copy, Clone, FromBytes, Debug)]
#[repr(C)]
pub struct GraphTileHeader {
    bit_field_1: FirstBitfield,
    base_lon_lat: [f32; 2],
    version: [u8; 16],
    /// The dataset ID (canonically, the last OSM changeset ID).
    pub dataset_id: u64,
    bit_field_2: SecondBitfield,
    transition_count_bitfield: TransitionCountBitfield,
    turn_lane_count_bitfield: TurnLaneCountBitfield,
    transit_record_bitfield: TransitRecordBitfield,
    misc_counts_bit_field_one: MiscCountsBitFieldOne,
    misc_counts_bit_field_two: MiscCountsBitFieldTwo,
    /// These can be used for custom information.
    /// As long as the size of the header and order of the date of the structure doesn't change,
    /// this should be backward compatible.
    pub reserved: u128,
    // TODO: Valhalla has the unhelpful comment "Offset to the beginning of the variable sized data."
    // We should improve these.
    pub complex_restriction_forward_offset: u32,
    pub complex_restriction_reverse_offset: u32,
    pub edge_info_offset: u32,
    pub text_list_offset: u32,
    /// The date the tile was created (since pivot date).
    /// TODO: Figure out what pivot date means; probably hide the raw value
    pub create_date: u32,
    /// Offsets for each bin of the grid (for search/lookup)
    pub bin_offsets: [u32; BIN_COUNT as usize],
    /// Offset to the beginning of the variable sized data.
    pub lane_connectivity_offset: u32,
    /// Offset to the beginning of the variable sized data.
    pub predicted_speeds_offset: u32,
    /// The size of the tile (in bytes)
    pub tile_size: u32,
    /// See valhalla/balder/graphtileheader.h for details.
    _empty_slots: [u32; EMPTY_SLOTS],
}

impl GraphTileHeader {
    /// The full Graph ID of this tile.
    #[inline]
    pub const fn graph_id(&self) -> GraphId {
        // Safety: We know that the bit field cannot contain a value
        // larger than the max allowed value.
        unsafe { GraphId::from_id_unchecked(self.bit_field_1.graph_id()) }
    }

    /// The relative road density within this tile (0-15).
    #[inline]
    pub const fn density(&self) -> u8 {
        self.bit_field_1.density()
    }

    /// The relative quality of name assignment for this tile (0-15).
    #[inline]
    pub const fn name_quality(&self) -> u8 {
        self.bit_field_1.name_quality()
    }

    /// The relative quality of speed assignment for this tile (0-15).
    #[inline]
    pub const fn speed_quality(&self) -> u8 {
        self.bit_field_1.speed_quality()
    }

    /// The relative quality of exit signs for this tile (0-15).
    #[inline]
    pub const fn exit_quality(&self) -> u8 {
        self.bit_field_1.exit_quality()
    }

    /// Does this tile include elevation data?
    #[inline]
    pub const fn has_elevation(&self) -> bool {
        self.bit_field_1.has_elevation() != 0
    }

    /// Does this tile include extended directed edge attributes?
    #[inline]
    pub const fn has_ext_directed_edge(&self) -> bool {
        self.bit_field_1.has_ext_directed_edge() != 0
    }

    /// The coordinate of the southwest corner of this graph tile.
    #[inline]
    pub const fn sw_corner(&self) -> Coord<f32> {
        coord! {x: self.base_lon_lat[0], y: self.base_lon_lat[1]}
    }

    /// Gets the version of Baldr used to generate this graph tile.
    pub fn version(&self) -> Cow<'_, str> {
        self.version.as_cow_str()
    }

    #[inline]
    pub const fn node_count(&self) -> u32 {
        self.bit_field_2.node_count()
    }

    /// The number of directed edges in this graph tile.
    #[inline]
    pub const fn directed_edge_count(&self) -> u32 {
        self.bit_field_2.directed_edge_count()
    }

    /// The number of predicted speed records in this graph tile.
    #[inline]
    pub const fn predicted_speeds_count(&self) -> u32 {
        self.bit_field_2.predicted_speeds_count()
    }

    /// The number of node transitions in this graph tile.
    #[inline]
    pub const fn transition_count(&self) -> u32 {
        self.transition_count_bitfield.transition_count()
    }

    /// The number of turn lanes in this graph tile.
    #[inline]
    pub const fn turn_lane_count(&self) -> u32 {
        self.turn_lane_count_bitfield.turn_lane_count()
    }

    /// The number of transit transfers in this graph tile.
    #[inline]
    pub const fn transfer_count(&self) -> u32 {
        self.transit_record_bitfield.transfer_count()
    }

    /// The number of transit departures in this graph tile.
    #[inline]
    pub const fn departure_count(&self) -> u32 {
        self.transit_record_bitfield.departure_count()
    }

    /// The number of transit stops in this graph tile.
    #[inline]
    pub const fn stop_count(&self) -> u16 {
        self.transit_record_bitfield.stop_count()
    }

    /// The number of transit routes in this graph tile.
    #[inline]
    pub const fn route_count(&self) -> u16 {
        self.misc_counts_bit_field_one.route_count()
    }

    /// The number of transit schedules in this graph tile.
    #[inline]
    pub const fn schedule_count(&self) -> u16 {
        self.misc_counts_bit_field_one.schedule_count()
    }

    /// The number of signs in this graph tile.
    #[inline]
    pub const fn sign_count(&self) -> u32 {
        self.misc_counts_bit_field_one.sign_count()
    }

    /// The number of access restrictions in this graph tile.
    #[inline]
    pub const fn access_restriction_count(&self) -> u32 {
        self.misc_counts_bit_field_two.access_restriction_count()
    }

    /// The number of admin records in this graph tile.
    #[inline]
    pub const fn admin_count(&self) -> u16 {
        self.misc_counts_bit_field_two.admin_count()
    }
}

#[cfg(test)]
mod test {
    use crate::graph_tile::TEST_GRAPH_TILE;

    #[test]
    fn test_parse_header() {
        let tile = &*TEST_GRAPH_TILE;

        insta::assert_debug_snapshot!(tile.header);
    }
}
