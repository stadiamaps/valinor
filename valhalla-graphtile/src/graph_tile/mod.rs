use thiserror::Error;
use zerocopy::{transmute, try_transmute};

use bytes::Bytes;
use enumset::EnumSet;
#[cfg(test)]
use std::sync::LazyLock;

// To keep files manageable, we will keep internal modules specific to each mega-struct,
// and publicly re-export the public type.
// Each struct should be tested on fixture tiles to a reasonable level of confidence.
// Leverage snapshot tests where possible, compare with Valhalla C++ results,
// and test the first and last elements of variable length vectors
// with snapshot tests.

mod access_restriction;
mod admin;
mod directed_edge;
mod edge_info;
mod header;
mod node;
mod sign;
mod transit;
mod turn_lane;

use crate::{
    graph_id::{GraphId, InvalidGraphIdError},
    transmute_variable_length_data, try_transmute_variable_length_data, Access,
};
pub use access_restriction::{AccessRestriction, AccessRestrictionType};
pub use admin::Admin;
pub use directed_edge::{DirectedEdge, DirectedEdgeExt};
pub use edge_info::EdgeInfo;
pub use header::GraphTileHeader;
pub use node::{NodeInfo, NodeTransition};
pub use sign::{Sign, SignType};
pub use transit::{TransitDeparture, TransitRoute, TransitSchedule, TransitStop, TransitTransfer};
pub use turn_lane::TurnLane;

#[derive(Debug, Error)]
pub enum GraphTileError {
    #[error("Unable to extract a slice of the correct length; the tile data is malformed.")]
    SliceArrayConversion(#[from] std::array::TryFromSliceError),
    #[error("The byte sequence is not valid for this type.")]
    ValidityError,
    #[error("Invalid graph ID.")]
    GraphIdParseError(#[from] InvalidGraphIdError),
}

#[derive(Debug, Error)]
pub enum LookupError {
    #[error("Mismatched base; the graph ID cannot exist in this tile.")]
    MismatchedBase,
    #[error("The feature at the index specified does not exist in this tile.")]
    InvalidIndex,
}

/// A tile within the Valhalla hierarchical tile graph.
pub struct GraphTile {
    memory: Bytes,
    /// Header with various metadata about the tile and internal sizes.
    pub header: GraphTileHeader,
    /// The list of nodes in the graph tile.
    nodes: Vec<NodeInfo>,
    /// The list of transitions between nodes on different levels.
    transitions: Vec<NodeTransition>,
    directed_edges: Vec<DirectedEdge>,
    ext_directed_edges: Vec<DirectedEdgeExt>,
    access_restrictions: Vec<AccessRestriction>,
    transit_departures: Vec<TransitDeparture>,
    transit_stops: Vec<TransitStop>,
    transit_routes: Vec<TransitRoute>,
    transit_schedules: Vec<TransitSchedule>,
    transit_transfers: Vec<TransitTransfer>,
    signs: Vec<Sign>,
    turn_lanes: Vec<TurnLane>,
    admins: Vec<Admin>,
    edge_bins: GraphId,
    // TODO: Complex forward restrictions
    // TODO: Complex reverse restrictions
    // TODO: Street names (names here)
    // TODO: Lane connectivity
    // TODO: Predicted speeds
    // TODO: Stop one stops(?)
    // TODO: Route one stops(?)
    // TODO: Operator one stops(?)
}

impl GraphTile {
    /// Gets the Graph ID of the tile.
    #[inline]
    pub fn graph_id(&self) -> GraphId {
        self.header.graph_id()
    }

    /// Does the supplied graph ID belong in this tile?
    ///
    /// A true result does not necessarily guarantee that an object with this ID exists,
    /// but in that case, either you've cooked up an invalid ID for fun,
    /// or the graph is invalid.
    pub fn may_contain_id(&self, id: &GraphId) -> bool {
        id.tile_base_id() == self.graph_id().tile_base_id()
    }

    /// Gets a reference to a node in this tile with the given graph ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the graph ID cannot be contained in this tile
    /// or the index is invalid.
    pub fn get_node(&self, id: &GraphId) -> Result<&NodeInfo, LookupError> {
        if self.may_contain_id(id) {
            self.nodes
                .get(id.index() as usize)
                .ok_or(LookupError::InvalidIndex)
        } else {
            Err(LookupError::MismatchedBase)
        }
    }

    /// Gets a reference to the directed edge in this tile by graph ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the graph ID cannot be contained in this tile
    /// or the index is invalid.
    pub fn get_directed_edge(&self, id: &GraphId) -> Result<&DirectedEdge, LookupError> {
        if self.may_contain_id(id) {
            self.directed_edges
                .get(id.index() as usize)
                .ok_or(LookupError::InvalidIndex)
        } else {
            Err(LookupError::MismatchedBase)
        }
    }

    /// Gets the opposing edge index, if and only if it exists in this tile.
    pub fn get_opp_edge_index(&self, graph_id: &GraphId) -> Result<u32, LookupError> {
        let edge = self.get_directed_edge(graph_id)?;

        // The edge might leave the tile, so we have to do a complicated lookup
        let end_node_id = edge.end_node_id();
        let opp_edge_index = edge.opposing_edge_index();

        // TODO: Probably a cleaner pattern here?
        let node_edge_index = self.get_node(&end_node_id).map(|n| n.edge_index())?;

        Ok(node_edge_index + opp_edge_index)
    }

    /// Gets a reference to an extended directed edge in this tile by graph ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the graph ID cannot be contained in this tile
    /// or the index is invalid.
    pub fn get_ext_directed_edge(&self, id: &GraphId) -> Result<&DirectedEdgeExt, LookupError> {
        if self.may_contain_id(id) {
            self.ext_directed_edges
                .get(id.index() as usize)
                .ok_or(LookupError::InvalidIndex)
        } else {
            Err(LookupError::MismatchedBase)
        }
    }

    /// Gets access restriction for a directed edge
    /// which apply to *any* of the supplied access modes (ex: auto, bicycle, etc.).
    pub fn get_access_restrictions(
        &self,
        directed_edge_index: u32,
        access_modes: EnumSet<Access>,
    ) -> Vec<&AccessRestriction> {
        // Find the point such that all elements below index are less than directed_edge_index
        // (NOTE: the list is pre-sorted during tile building)
        let index = self
            .access_restrictions
            .partition_point(|e| e.edge_index() < directed_edge_index);

        self.access_restrictions
            .iter()
            // Iterate over the relevant objects (this is an empty iterator if index >= length)
            .skip(index)
            .take_while(|e| e.edge_index() == directed_edge_index)
            .filter(|e| !e.affected_access_modes().is_disjoint(access_modes))
            .collect()
    }

    /// Gets edge info for a directed edge.
    pub fn get_edge_info(&self, directed_edge: &DirectedEdge) -> Result<EdgeInfo, GraphTileError> {
        let edge_info_start = self.header.edge_info_offset as usize;
        let edge_info_offset = directed_edge.edge_info_offset() as usize;

        // TODO: Text slice as a separate property or a function?
        let text_start = self.header.text_list_offset as usize;
        let text_size = self.header.lane_connectivity_offset as usize - text_start;

        Ok(EdgeInfo::try_from((
            self.memory.slice(edge_info_start + edge_info_offset..),
            self.memory.slice(text_start..text_start + text_size),
        ))?)
    }
}

// TODO: Feels like this could be a macro
impl TryFrom<Bytes> for GraphTile {
    type Error = GraphTileError;

    fn try_from(bytes: Bytes) -> Result<Self, Self::Error> {
        let value = bytes.as_ref();
        // Get the byte range of the header so we can transmute it
        const HEADER_SIZE: usize = size_of::<GraphTileHeader>();
        let header_range = 0..HEADER_SIZE;

        // Save the pointer to the list of nodes (the range is consumed)
        let offset = header_range.end;

        // Header
        let header_slice: [u8; HEADER_SIZE] = value[header_range].try_into()?;
        let header: GraphTileHeader = transmute!(header_slice);

        // All the variably sized data arrays

        let (nodes, offset) =
            transmute_variable_length_data!(NodeInfo, value, offset, header.node_count() as usize)?;

        let (transitions, offset) = transmute_variable_length_data!(
            NodeTransition,
            value,
            offset,
            header.transition_count() as usize
        )?;

        let (directed_edges, offset) = try_transmute_variable_length_data!(
            DirectedEdge,
            value,
            offset,
            header.directed_edge_count() as usize
        )?;

        let directed_edge_ext_count = if header.has_ext_directed_edge() {
            header.directed_edge_count() as usize
        } else {
            0
        };
        let (ext_directed_edges, offset) = transmute_variable_length_data!(
            DirectedEdgeExt,
            value,
            offset,
            directed_edge_ext_count
        )?;

        let (access_restrictions, offset) = try_transmute_variable_length_data!(
            AccessRestriction,
            value,
            offset,
            header.access_restriction_count() as usize
        )?;

        let (transit_departures, offset) = transmute_variable_length_data!(
            TransitDeparture,
            value,
            offset,
            header.departure_count() as usize
        )?;

        let (transit_stops, offset) = transmute_variable_length_data!(
            TransitStop,
            value,
            offset,
            header.stop_count() as usize
        )?;

        let (transit_routes, offset) = transmute_variable_length_data!(
            TransitRoute,
            value,
            offset,
            header.route_count() as usize
        )?;

        let (transit_schedules, offset) = transmute_variable_length_data!(
            TransitSchedule,
            value,
            offset,
            header.schedule_count() as usize
        )?;

        let (transit_transfers, offset) = transmute_variable_length_data!(
            TransitTransfer,
            value,
            offset,
            header.transfer_count() as usize
        )?;

        let (signs, offset) =
            try_transmute_variable_length_data!(Sign, value, offset, header.sign_count() as usize)?;

        let (turn_lanes, offset) = try_transmute_variable_length_data!(
            TurnLane,
            value,
            offset,
            header.turn_lane_count() as usize
        )?;

        let (admins, offset) = try_transmute_variable_length_data!(
            Admin,
            value,
            offset,
            header.admin_count() as usize
        )?;

        const U64_SIZE: usize = size_of::<u64>();
        let slice: [u8; U64_SIZE] = (&value[offset..offset + U64_SIZE]).try_into()?;
        let raw_graph_id = transmute!(slice);
        let edge_bins = GraphId::try_from_id(raw_graph_id)?;

        Ok(Self {
            memory: bytes,
            header,
            nodes,
            transitions,
            directed_edges,
            ext_directed_edges,
            access_restrictions,
            transit_departures,
            transit_stops,
            transit_routes,
            transit_schedules,
            transit_transfers,
            signs,
            turn_lanes,
            admins,
            edge_bins,
        })
    }
}

#[cfg(test)]
const TEST_GRAPH_TILE_ID: GraphId = unsafe { GraphId::from_components_unchecked(0, 3015, 0) };

#[cfg(test)]
static TEST_GRAPH_TILE: LazyLock<GraphTile> = LazyLock::new(|| {
    let relative_path = TEST_GRAPH_TILE_ID
        .file_path("gph")
        .expect("Unable to get relative path");
    let path = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("fixtures")
        .join("andorra-tiles")
        .join(relative_path);
    let bytes = Bytes::from(std::fs::read(path).expect("Unable to read file"));

    GraphTile::try_from(bytes).expect("Unable to get tile")
});

#[cfg(test)]
mod tests {
    use crate::graph_tile::{TEST_GRAPH_TILE, TEST_GRAPH_TILE_ID};
    use crate::Access;
    use enumset::{enum_set, EnumSet};

    #[test]
    fn test_get_opp_edge_index() {
        let tile = &*TEST_GRAPH_TILE;

        let edges: Vec<_> = (0..u64::from(tile.header.directed_edge_count()))
            .map(|index| {
                let edge_id = TEST_GRAPH_TILE_ID
                    .with_index(index)
                    .expect("Invalid graph ID.");
                let opp_edge_index = tile
                    .get_opp_edge_index(&edge_id)
                    .expect("Unable to get opp edge index.");
                opp_edge_index
            })
            .collect();
        insta::assert_debug_snapshot!(edges);
    }

    #[test]
    fn test_get_access_restrictions() {
        let tile = &*TEST_GRAPH_TILE;

        let mut found_restriction_count = 0;
        let mut found_partial_subset_count = 0;
        for edge_index in 0..tile.header.directed_edge_count() {
            // Sanity check edge index with no access mode filter
            let all_edge_restrictions = tile.get_access_restrictions(edge_index, EnumSet::all());
            for restriction in &all_edge_restrictions {
                found_restriction_count += 1;
                assert_eq!(restriction.edge_index(), edge_index);
            }

            // In the tile fixture, all restrictions apply to trucks
            assert_eq!(
                all_edge_restrictions,
                tile.get_access_restrictions(edge_index, enum_set!(Access::Truck))
            );

            // Empty set should yield no access restrictions
            assert!(tile
                .get_access_restrictions(edge_index, EnumSet::empty())
                .is_empty());

            // This set includes one mode that matches 4 elements in the fixture tile,
            // and another mode that matches none.
            for restriction in
                tile.get_access_restrictions(edge_index, enum_set!(Access::Auto | Access::GolfCart))
            {
                found_partial_subset_count += 1;
                assert_eq!(restriction.edge_index(), edge_index);
            }
        }

        // Hard-coded bits based on the test fixture
        assert_eq!(found_restriction_count, 8);
        assert_eq!(found_partial_subset_count, 4);
    }

    #[test]
    fn test_edge_bins() {
        // TODO: TBH I don't actually understand how this is a valid graph tile. Clearly some black bit magic.
        let tile = &*TEST_GRAPH_TILE;
        assert_eq!(tile.edge_bins.level(), 0);
        assert_eq!(tile.edge_bins.tile_id(), 0);
        assert_eq!(tile.edge_bins.index(), 32000);
    }

    #[test]
    fn test_edge_info() {
        let tile = &*TEST_GRAPH_TILE;
        let first_edge_info = tile
            .get_edge_info(&tile.directed_edges[0])
            .expect("Unable to get edge info.");
        insta::assert_debug_snapshot!("first_edge_info", first_edge_info);
        insta::assert_debug_snapshot!("first_edge_info_decoded_shape", first_edge_info.shape());
        insta::assert_debug_snapshot!("first_edge_info_names", first_edge_info.get_names());

        let last_edge_info = tile
            .get_edge_info(&tile.directed_edges.last().unwrap())
            .expect("Unable to get edge info.");
        insta::assert_debug_snapshot!("last_edge_info", last_edge_info);

        // Edge chosen somewhat randomly; it happens to have multiple names.
        let other_edge_info = tile
            .get_edge_info(&tile.directed_edges[2000])
            .expect("Unable to get edge info.");
        insta::assert_debug_snapshot!("other_edge_info", other_edge_info);
        insta::assert_debug_snapshot!("other_edge_info_names", other_edge_info.get_names());
    }
}
