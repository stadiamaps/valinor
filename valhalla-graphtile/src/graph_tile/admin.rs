use crate::AsCowStr;
use std::borrow::Cow;
use zerocopy_derive::FromBytes;

#[derive(FromBytes, Debug)]
#[repr(C)]
pub struct Admin {
    /// The offset into the [`GraphTile`](super::GraphTile) text list for the country name.
    pub country_name_offset: u32,
    /// The offset into the [`GraphTile`](super::GraphTile) text list for the principal subdivision name.
    pub principal_subdivision_offset: u32,
    country_iso: [u8; 2],
    principal_subdivision_iso: [u8; 3],
    // TODO: Do we need 3 bytes for alignment?
}

impl Admin {
    /// Gets the ISO 3166-1 country code
    pub fn country_iso(&self) -> Cow<'_, str> {
        self.country_iso.as_cow_str()
    }

    /// Gets the ISO 3166-2 principal subdivision code
    pub fn principal_subdivision_iso(&self) -> Cow<'_, str> {
        self.principal_subdivision_iso.as_cow_str()
    }
}

#[cfg(test)]
mod tests {
    use crate::graph_tile::TEST_GRAPH_TILE;

    #[test]
    fn test_parse_admin_count() {
        let tile = &*TEST_GRAPH_TILE;

        assert_eq!(tile.admins.len(), tile.header.admin_count() as usize);
    }

    #[test]
    fn test_parse_admins() {
        let tile = &*TEST_GRAPH_TILE;

        insta::assert_debug_snapshot!(tile.admins);
        insta::assert_debug_snapshot!(tile
            .admins
            .iter()
            .map(|admin| admin.country_iso())
            .collect::<Vec<_>>());
        insta::assert_debug_snapshot!(tile
            .admins
            .iter()
            .map(|admin| admin.principal_subdivision_iso())
            .collect::<Vec<_>>());
    }
}
