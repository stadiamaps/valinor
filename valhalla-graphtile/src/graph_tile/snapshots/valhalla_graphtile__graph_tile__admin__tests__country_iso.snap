---
source: valhalla-graphtile/src/graph_tile/admin.rs
expression: "tile.admins.iter().map(|admin| admin.country_iso()).collect::<Vec<_>>()"
---
[
    "",
    "AD",
    "ES",
    "FR",
]
